x/gamm/keeper/migrate.go

```diff
@@ -13,56 +13,44 @@ import (
+func (k Keeper) Migrate(ctx sdk.Context, sender sdk.AccAddress, sharesToMigrate sdk.Coin) (amount0, amount1 sdk.Int, liquidity sdk.Dec, poolIdLeaving, poolIdEntering uint64, err error) {
 	// Get the balancer poolId by parsing the gamm share denom.
-	poolIdLeaving, err = getPoolIdFromSharesDenom(sharesToMigrate.Denom)
-	if err != nil {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
-	}
+	poolIdLeaving = types.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
 
-	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
-	migrationInfo := k.GetMigrationInfo(ctx)
-	matchFound := false
-	for _, info := range migrationInfo.BalancerToConcentratedPoolLinks {
-		if info.BalancerPoolId == poolIdLeaving {
-			if info.ClPoolId != poolIdEntering {
-				return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, types.InvalidPoolMigrationLinkError{PoolIdEntering: poolIdEntering, CanonicalId: info.ClPoolId}
-			}
-			matchFound = true
-			break
-		}
-	}
-	if !matchFound {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, types.PoolMigrationLinkNotFoundError{PoolIdLeaving: poolIdLeaving}
+	// Find the governance sanctioned link between the balancer pool and a concentrated pool.
+	poolIdEntering, err = k.GetLinkedConcentratedPool(ctx, poolIdLeaving)
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, err
 	}
 
 	// Get the concentrated pool from the message and type cast it to ConcentratedPoolExtension.
 	poolI, err := k.clKeeper.GetPool(ctx, poolIdEntering)
 	if err != nil {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, err
 	}
 	concentratedPool, ok := poolI.(cltypes.ConcentratedPoolExtension)
 	if !ok {
 		// If the conversion fails, return an error.
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, fmt.Errorf("given pool does not implement ConcentratedPoolExtension, implements %T", poolI)
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, fmt.Errorf("given pool does not implement ConcentratedPoolExtension, implements %T", poolI)
 	}
 
 	// Exit the balancer pool position.
 	exitCoins, err := k.ExitPool(ctx, sender, poolIdLeaving, sharesToMigrate.Amount, sdk.NewCoins())
 	if err != nil {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, err
+		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, 0, 0, err
 	}
 
 	// Determine the max and min ticks for the concentrated pool we are migrating to.
 	minTick, maxTick := cl.GetMinAndMaxTicksFromExponentAtPriceOne(concentratedPool.GetPrecisionFactorAtPriceOne())
 
 	// Create a full range (min to max tick) concentrated liquidity position.
-	// TODO: Will need to implement lock breaking logic and add the corresponding freeze duration here.
 	amount0, amount1, liquidity, err = k.clKeeper.CreatePosition(ctx, poolIdEntering, sender, exitCoins.AmountOf(concentratedPool.GetToken0()), exitCoins.AmountOf(concentratedPool.GetToken1()), sdk.ZeroInt(), sdk.ZeroInt(), minTick, maxTick, time.Time{})
```

Comment:
Note to reviewer: We freeze for the zero value of time here since only unlocked LP tokens can get migrated through this entry point.